crates/bevy_reflect/src/path.rs

```diff
@@ -104,270 +114,481 @@ impl<T: Reflect> GetPath for T {
+                .field_at_mut(*field_index)
+                .ok_or(ReflectPathError::InvalidFieldIndex {
                     index: current_index,
-                    tuple_struct_index: tuple_index,
-                },
-            )?)
+                    field_index: *field_index,
+                }),
+            (Self::TupleIndex(tuple_index), ReflectMut::TupleStruct(reflect_struct)) => {
+                reflect_struct.field_mut(*tuple_index).ok_or(
+                    ReflectPathError::InvalidTupleStructIndex {
+                        index: current_index,
+                        tuple_struct_index: *tuple_index,
+                    },
+                )
+            }
+            (Self::ListIndex(list_index), ReflectMut::List(reflect_list)) => reflect_list
+                .get_mut(*list_index)
+                .ok_or(ReflectPathError::InvalidListIndex {
+                    index: current_index,
+                    list_index: *list_index,
+                }),
+            (Self::ListIndex(_), _) => Err(ReflectPathError::ExpectedList {
+                index: current_index,
+            }),
+            _ => Err(ReflectPathError::ExpectedStruct {
+                index: current_index,
+            }),
         }
-        _ => Err(ReflectPathError::ExpectedStruct {
-            index: current_index,
-        }),
     }
 }
 
-enum Token<'a> {
-    Dot,
-    OpenBracket,
-    CloseBracket,
-    Ident(&'a str),
+struct PathParser<'a> {
+    path: &'a str,
+    index: usize,
 }
 
-fn next_token<'a>(path: &'a str, index: &mut usize) -> Option<Token<'a>> {
-    if *index >= path.len() {
-        return None;
+impl<'a> PathParser<'a> {
+    fn new(path: &'a str) -> Self {
+        Self { path, index: 0 }
     }
 
-    match path[*index..].chars().next().unwrap() {
-        '.' => {
-            *index += 1;
-            return Some(Token::Dot);
+    fn next_token(&mut self) -> Option<Token<'a>> {
+        if self.index >= self.path.len() {
+            return None;
         }
-        '[' => {
-            *index += 1;
-            return Some(Token::OpenBracket);
+
+        match self.path[self.index..].chars().next().unwrap() {
+            '.' => {
+                self.index += 1;
+                return Some(Token::Dot);
+            }
+            '#' => {
+                self.index += 1;
+                return Some(Token::CrossHatch);
+            }
+            '[' => {
+                self.index += 1;
+                return Some(Token::OpenBracket);
+            }
+            ']' => {
+                self.index += 1;
+                return Some(Token::CloseBracket);
+            }
+            _ => {}
         }
-        ']' => {
-            *index += 1;
-            return Some(Token::CloseBracket);
+
+        // we can assume we are parsing an ident now
+        for (char_index, character) in self.path[self.index..].chars().enumerate() {
+            match character {
+                '.' | '#' | '[' | ']' => {
```

Comment:
This isn't that important, but since you're working on the file, maybe we can turn these into `const`s. Just to prevent accidental issues for future contributors and to organize all expected tokens.